src/cascade/model/model_writer.py

```diff
@@ -0,0 +1,356 @@
+        unique_times = self._times[np.unique(self._times.round(decimals=14), return_index=True)[1]]
+        unique_times.sort()
+        if unique_times[-1] - unique_times[0] < 1:
+            unique_times = np.append(unique_times, unique_times[-1] + 1)
+        self._dismod_file.time = pd.DataFrame(dict(time_id=range(len(unique_times)), time=unique_times))
+        self._flushed = True
+
+    def _integrand_id_func(self, name):
+        return int(self._dismod_file.integrand.query("integrand_name==@name").integrand_id.values[0])
+
+    def _rate_id_func(self, name):
+        return int(self._dismod_file.rate.query("rate_name==@name").rate_id)
+
+    def _fix_ages_times(self, df):
+        """Given a Pandas df with age and time columns, assign age_id and time_id
+        columns using a nearest match. Keep the same order. Allow nan ages
+        or times."""
+        assert "age" in df.columns
+        assert "time" in df.columns
+        df = df.assign(save_idx=df.index)
+        for dat in ["age", "time"]:
+            col_id = f"{dat}_id"
+            sort_by = df.sort_values(dat)
+            in_grid = sort_by[dat].notna()
+            at_table = getattr(self._dismod_file, dat)
+            aged = pd.merge_asof(sort_by[in_grid], at_table, on=dat, direction="nearest")
+            df = df.merge(aged[["save_idx", col_id]], on="save_idx", how="left")
+        assert "age_id" in df.columns
+        assert "time_id" in df.columns
+        return df.drop("save_idx", axis=1)
+
+    def write_covariate(self, covariates):
+        self._dismod_file.covariate = self._dismod_file.empty_table("covariate")
+        reorder = list()
+        lookup = {search.name: search for search in covariates}
+        for special in ["sex", "one"]:
+            if special in lookup:
+                reorder.append(lookup[special])
+                del lookup[special]
+        for remaining in sorted(lookup.keys()):
+            reorder.append(lookup[remaining])
+        CODELOG.debug(f"covariates {', '.join(c.name for c in reorder)}")
+        covariate_df, cov_col_id_func, covariate_renames = _make_covariate_table(reorder)
+        self._dismod_file.update_table_columns("covariate", covariate_df)
+        self._dismod_file.covariate = covariate_df
+        self._covariate_id_func = cov_col_id_func
+        self._session.covariate_rename = covariate_renames
+
+
+def _make_covariate_table(covariates):
+    null_references = list()
+    for check_ref_col in covariates:
+        if not isinstance(check_ref_col.reference, Real):
+            null_references.append(check_ref_col.name)
+    if null_references:
+        raise RuntimeError(f"Covariate columns without reference values {null_references}")
+
+    # Dismod requires us to rename covariates from names like sex, and "one"
+    # to x_0, x_1,... They must be "x_<digit>".
+    renames = dict()
+    for cov_idx, covariate in enumerate(covariates):
```

Comment:
Sort? It would be nice if what we write is stable relative to minor perturbations. I haven't actually been watching for that. I'm actually not sure how important it really is. But it feels like something we'll miss eventually. Or maybe no human will ever have to look at one of these sqlite files ever again and that would be fine.

Reply:
They are sorted one level above this function. Always sex, then "one" covariate, and then the rest in sorted order. Very carefully. Greg still has his ways to debug the db files, and I'm trying to keep those working.